### DISCUSSION TITLE
How to implement Auto-DevOps (Auto CI/CD with AWS CFT) for pimcore?

### DISCUSSION THREAD

**Discussion by kevindesai** [pimcore/pimcore#10413] [Q&A / Support]
Hello,
I am implementing Auto DevOps for pimcore application where the goal is the auto deploy the application & DB on several AWS Services. 

The application will be deployed in AWS ECS using Docker and ECR and the database will be deployed on AWS Aurora. Also as part of this CI/CD the infrastructure will be auto created using AWS CFT.

There are two separate pipeline for Infrastructure and application
1) Gitlab Pipeline for Application deployment
2) Gitlab pipeline for infrastructure building using AWS CFT

Infrastructure pipeline will be used one time for each environment setup (e.g. Dev, QA, Stage and Production).
For example when the QA infra needs to the setup,  the infrastructure pipeline will be used to create resources in AWS like ECS and Aurora DB cluster

Application pipeline will be triggered on every new build (i.e. when the application code is pushed/merged to a respective gitlab repo/branch). 

I am facing below challenge regarding pimcore-install command which I want to use in the application deployment pipeline.

1. The Pimcore Database is required to be on a private-subnet (i.e. cannot be directly accessible outside the VPC or via the internet).

2. The Pimcore Database should be initialized only once at the time the database is first created. This means the database should not be initialized again when a new docker Image is created when the source of the PIM application is updated.

3. Since the pimcore-install command is part of the application deployment pipeline, it initializes the database on every build.

How can I solve the problem where the pimcore-install command is configured in the application deployment pipeline such a way that when the deployment is happening for the very first time, it initializes the database but for any subsequent deployment, the database is not altered in any way.

I want to automate this and don’t want to alter the pipelines manually or trigger the command manually the first time.

Is there a guide that I can follow that covers a similar use-case?

**Comment by RajatToshniwal** [comment]
I believe the easiest hack is to create a shell wrapper around Pimcore-install command and check for tables like object in the database, whether it exist or not. Otherwise you can also check the presence of database.yaml file in the code directory depending upon what all you are committing in the code repo.

BTW, there are other approaches as well depending upon what business case you want to solve like you can completely avoid the Pimcore-install.

- You would like to change your infrastructure CFT to create the MySQL database from snapshot instead of newly creating it. Your snapshot consist of raw pimcore database. You can have many snapshots based on the version of database and Pimcore.
- You can also execute lambda or command runner feature of CFT after the execution of your infrastructure CFT to create entry on s3 or using some external API for the newly created Aurora and then you can remove the entry during your first run.
- If the only requirement is to procure the database using sql dump file, you can instantiate EC2 instance in the Public subnet of your VPC on temporary basis and mysql restore will sort our your issue.

Let me know, if that suffice your need.

**Reply by macentric** [reply]
Hi Kevin and Rajat ,
Hope you doing well, i'm new to devops and so to pimcore could you please guide me the right path to deploy ci/cd for pimcore on selfhosted gitlab 
thanks in advance
